rust/src/acl/mod.rs

```diff
@@ -0,0 +1,100 @@
+// Copyright 2023 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! An implementation of Matrix server ACL rules.
+
+use crate::push::utils::{glob_to_regex, GlobMatchType};
+use anyhow::Error;
+use pyo3::prelude::*;
+use regex::Regex;
+use std::net::Ipv4Addr;
+use std::str::FromStr;
+
+/// Called when registering modules with python.
+pub fn register_module(py: Python<'_>, m: &PyModule) -> PyResult<()> {
+    let child_module = PyModule::new(py, "acl")?;
+    child_module.add_class::<ServerAclEvaluator>()?;
+
+    m.add_submodule(child_module)?;
+
+    // We need to manually add the module to sys.modules to make `from
+    // synapse.synapse_rust import acl` work.
+    py.import("sys")?
+        .getattr("modules")?
+        .set_item("synapse.synapse_rust.acl", child_module)?;
+
+    Ok(())
+}
+
+#[derive(Debug, Clone)]
+#[pyclass(frozen)]
+pub struct ServerAclEvaluator {
+    allow_ip_literals: bool,
+    allow: Vec<Regex>,
+    deny: Vec<Regex>,
+}
+
+#[pymethods]
+impl ServerAclEvaluator {
+    #[new]
+    pub fn py_new(
+        allow_ip_literals: bool,
+        allow: Vec<String>,
+        deny: Vec<String>,
```

Comment:
I was gonna say "should this be something like `IntoIterator<&str>`... but maybe that's not important because Python is going to give us `List[str]`, and FFI converts this to a `Vec<String>`?

Reply:
Pretty much pyo3 converts `List[str]` to `Vec<String>`, see https://pyo3.rs/v0.17.1/conversions/tables

Reply:
If you wanted to avoid the extra allocations for each string, I *think* you can do `Vec<&str>`? (I think the relevant conversion functions are at https://docs.rs/pyo3/latest/src/pyo3/conversions/std/string.rs.html#113-137)

Reply:
No reason not to, I think?